## Applied Data Science Capstone Assignment 2 :Segmenting and Clustering Neighborhoods in Toronto - Part 1


### For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

1. Start by creating a new Notebook for this assignment.
2. Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below:


In [1]:
# Import necessary libraries

import requests
import lxml.html as lh
import pandas as pd

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

#Create a handle, page, to handle the contents of the website
page = requests.get(url)

#Store the contents of the website under doc
doc = lh.fromstring(page.content)

#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

In [3]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

This means that there are 3 columns per row

In [4]:
# Parse the first row as our header
tr_elements = doc.xpath('//tr')

#Create empty list
col=[]
i=0

#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ('%d:"%s"'%(i,name))
    col.append((name,[]))

1:"Postal Code
"
2:"Borough
"
3:"Neighbourhood
"


#### Creating Pandas DataFrame

Each header is appended to a tuple along with an empty list.

In [5]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

In [6]:
# Check the length of each column. Ideally, they should all be the same
[len(C) for (title,C) in col]

[181, 181, 181]

This shows that each of the 3 columns has exactly 181 rows

#### Creating the pandas data frame 

In [7]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

3. To create the above dataframe:

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.
4. Submit a link to your Notebook on your Github repository. (10 marks)

Note: There are different website scraping libraries and packages in Python. For scraping the above table, you can simply use pandas to read the table into a pandas dataframe.

Another way, which would help to learn for more complicated cases of web scraping is using the BeautifulSoup package. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/

Use pandas, or the BeautifulSoup package, or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.

In [12]:
# Access the top 12 rows of the data frame 
df = df.replace('\n',' ', regex=True)
df.head(12)

Postcode Borough      Neighbourhood
0                                  Not assigned     M1A       Not assigned 
1                                  Not assigned     M2A       Not assigned 
2                                     Parkwoods     M3A         North York 
3                              Victoria Village     M4A         North York 
4                     Regent Park, Harbourfront     M5A   Downtown Toronto 
5              Lawrence Manor, Lawrence Heights     M6A         North York 
6   Queen's Park, Ontario Provincial Government     M7A   Downtown Toronto 
7                                  Not assigned     M8A       Not assigned 
8       Islington Avenue, Humber Valley Village     M9A          Etobicoke 
9                                Malvern, Rouge     M1B        Scarborough 
10                                 Not assigned     M2B       Not assigned 
11                                    Don Mills     M3B         North York

#### Rearranging and renaming the columns

In [13]:
df.columns = ['Borough', 'Neighbourhood','Postcode']

cols = df.columns.tolist()
cols

cols = cols[-1:] + cols[:-1]

df = df[cols]

df = df.replace('\n',' ', regex=True)
df.head()

Postcode                     Borough Neighbourhood
0      Not assigned                Not assigned           M1A 
1      Not assigned                Not assigned           M2A 
2        North York                   Parkwoods           M3A 
3        North York            Victoria Village           M4A 
4  Downtown Toronto   Regent Park, Harbourfront           M5A

#### Cleaning the messy string in the Borough column

In [14]:
df = df.replace('\n',' ', regex=True)
df.head(12)

Postcode                                       Borough  \
0       Not assigned                                  Not assigned    
1       Not assigned                                  Not assigned    
2         North York                                     Parkwoods    
3         North York                              Victoria Village    
4   Downtown Toronto                     Regent Park, Harbourfront    
5         North York              Lawrence Manor, Lawrence Heights    
6   Downtown Toronto   Queen's Park, Ontario Provincial Government    
7       Not assigned                                  Not assigned    
8          Etobicoke       Islington Avenue, Humber Valley Village    
9        Scarborough                                Malvern, Rouge    
10      Not assigned                                  Not assigned    
11        North York                                     Don Mills    

   Neighbourhood  
0           M1A   
1           M2A   
2           M3A   
3           M4A   
4           M5A   
5           M6A   
6           M7A   
7           M8A   
8           M9A   
9           M1B   
10          M2B   
11          M3B

#### Dropping all cells with a borough that is Not assigned

In [15]:
df.drop(df.index[df['Borough'] == 'Not assigned'], inplace = True)

# Reset the index and dropping the previous index
df = df.reset_index(drop=True)

df.head(10)

Postcode                                       Borough  \
0      Not assigned                                  Not assigned    
1      Not assigned                                  Not assigned    
2        North York                                     Parkwoods    
3        North York                              Victoria Village    
4  Downtown Toronto                     Regent Park, Harbourfront    
5        North York              Lawrence Manor, Lawrence Heights    
6  Downtown Toronto   Queen's Park, Ontario Provincial Government    
7      Not assigned                                  Not assigned    
8         Etobicoke       Islington Avenue, Humber Valley Village    
9       Scarborough                                Malvern, Rouge    

  Neighbourhood  
0          M1A   
1          M2A   
2          M3A   
3          M4A   
4          M5A   
5          M6A   
6          M7A   
7          M8A   
8          M9A   
9          M1B

#### Combining Neighbourhoods based on similar Postcode and Borough  

In [20]:
df = df.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(','.join).reset_index()
df.columns = ['Postcode','Borough','Neighbourhood']
df.head(10)

Postcode                                            Borough  \
0  Canadian postal codes                                                       
1        Central Toronto                                         Davisville    
2        Central Toronto                                   Davisville North    
3        Central Toronto    Forest Hill North & West, Forest Hill Road Park    
4        Central Toronto                                      Lawrence Park    
5        Central Toronto                        Moore Park, Summerhill East    
6        Central Toronto                 North Toronto West,  Lawrence Park    
7        Central Toronto                                           Roselawn    
8        Central Toronto   Summerhill West, Rathnelly, South Hill, Forest...   
9        Central Toronto                The Annex, North Midtown, Yorkville    

  Neighbourhood  
0                
1           M4S  
2           M4P  
3           M5P  
4           M4N  
5           M4T  
6           M4R  
7           M5N  
8           M4V  
9           M5R

#### Removing any space in the start of the string

In [21]:
df['Neighbourhood'] = df['Neighbourhood'].str.strip()

#### Assigning Borough values to the Neignbourhood where vlaue is "Not assigned"

In [24]:
df.loc[df['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df['Borough']

In [25]:
# Check if the Neighbourhood for Queen's Park changed 
df[df['Borough'] == 'Queen\'s Park']

Empty DataFrame
Columns: [Postcode, Borough, Neighbourhood]
Index: []

In [26]:
# Check the shape of the data frame
df.shape

(101, 3)

#### Save this file to a csv

In [31]:
df.to_csv(r'df_can.csv')